cpp/mrc/src/internal/system/device_info.cpp

```diff
@@ -17,61 +17,163 @@
 
 #include "internal/system/device_info.hpp"
 
-#include "mrc/cuda/common.hpp"  // IWYU pragma: associated
+#include "mrc/cuda/common.hpp"
 
 #include <cuda_runtime.h>
+#include <dlfcn.h>
 #include <glog/logging.h>
+#include <hwloc.h>
+#include <hwloc/bitmap.h>
+#include <hwloc/linux.h>
 #include <nvml.h>
 
 #include <array>
+#include <cerrno>
 #include <cstddef>
+#include <cstdio>
+#include <memory>
 #include <ostream>
 #include <set>
+#include <stdexcept>
 #include <string>
 
+namespace mrc::internal::system {
+struct NvmlHandle;
+struct NvmlState;
+
 #define TEST_BIT(_n, _p) (_n & (1UL << _p))
 
-#define MRC_CHECK_NVML(expression)                                    \
-    {                                                                 \
-        auto status = (expression);                                   \
-        if (status != NVML_SUCCESS)                                   \
-        {                                                             \
-            LOG(FATAL) << "NVML failed: " << nvmlErrorString(status); \
-        }                                                             \
+#define MRC_CHECK_NVML(expression)                                                           \
+    {                                                                                        \
+        auto __status = (expression);                                                        \
+        if (__status != NVML_SUCCESS)                                                        \
+        {                                                                                    \
+            LOG(FATAL) << "NVML failed running '" << #expression                             \
+                       << "'. Error msg: " << NvmlState::handle().nvmlErrorString(__status); \
+        }                                                                                    \
+    }
+
+#define LOAD_NVTX_SYM(dll_ptr, function_var)                                                        \
+    function_var = reinterpret_cast<decltype(function_var)>(dlsym(dll_ptr, #function_var));         \
+    if (function_var == nullptr)                                                                    \
+    {                                                                                               \
+        throw std::runtime_error("Failed to load symbol " #function_var " from libnvidia-ml.so.1"); \
+    }
+
+/**
+ * @brief This class wraps all calls to NVML and ensures that the library is dynamically loaded at runtime using
+ * `dlopen`. This is necessary to remove the link dependency to `libnvidia-ml.so` which is a driver library. Because
+ * driver libraries are loaded into the docker container via the NVIDIA container runtime, we do not want to directly
+ * link to them. As well, if the driver is not found, we can still run MRC without GPUs. This gives us an oportunity to
+ * catch the missing library instead of the OS throwing an error.
+ *
+ */
+struct NvmlHandle
```

Comment:
Might make sense to have this be a singleton, or have a static counter to keep track of how many outstanding references to these symbols there are, which provides a way to check if anything is still using them when dlclose is called.